### PR TITLE
Add reverse code

### PR DESCRIPTION
**Pull request by BriungRi**

### CODE REVIEW

File: initial-reverse/reverse.cc [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+ *   Organization:
+ *
+ * =====================================================================================
+ */
+#include <iostream>
+#include <stdexcept>
+#include <sys/stat.h>
+
+template <class T> class Vector {
+  // Backing array
```
Comment (Juliang0705) [inline initial-reverse/reverse.cc:24]: put it in private section. private variables should have underscore at the end of their names. private section should also come after the public section

File: initial-reverse/reverse.cc [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+  int len;
+
+public:
+  // Constructor
+  Vector() {
+    cap = 1;
+    len = 0;
+
+    // Initialize a to point to an array of length `cap`
+    arr = (T *)malloc(cap * sizeof(T));
```
Comment (Juliang0705) [inline initial-reverse/reverse.cc]: use New. malloc is not type safe

File: initial-reverse/reverse.cc [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+  // Inserts `v` at the end
+  void insert(const T &v) {
+    if (len == cap) {
+      // Copy `arr` into a new array
+      cap *= 2;
+      T *nextArr = (T *)malloc(cap * sizeof(T));
+      for (int i = 0; i < len; i++) {
+        nextArr[i] = arr[i];
+      }
+      arr = nextArr;
```
Comment (Juliang0705) [inline initial-reverse/reverse.cc]: memory leak

File: initial-reverse/reverse.cc [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+
+    arr[len] = v;
+    len += 1;
+  }
+
+  // Gets the last element
+  T get() { return arr[len - 1]; }
+
+  // Gets the element at index `i`
+  T get(int i) {
```
Comment (Juliang0705) [inline initial-reverse/reverse.cc]: const int

File: initial-reverse/reverse.cc [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+      }
+      arr = nextArr;
+    }
+
+    arr[len] = v;
+    len += 1;
+  }
+
+  // Gets the last element
+  T get() { return arr[len - 1]; }
```
Comment (Juliang0705) [inline initial-reverse/reverse.cc]: do you want it to return a const reference instead? can the member function be marked const too?

File: initial-reverse/reverse.cc (hunk 4 above)
Comment (Juliang0705) [reply initial-reverse/reverse.cc]: also return a const reference instead?

File: initial-reverse/reverse.cc [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+
+  // Gets the element at index `i`
+  T get(int i) {
+    if (i < 0 || i >= len) {
+      char errMsg[100];
+      sprintf(errMsg, "Expected an index between 0 and %d, got %d", len - 1, i);
```
Comment (Juliang0705) [inline initial-reverse/reverse.cc]: use std::string

File: initial-reverse/reverse.cc [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+      throw std::out_of_range(errMsg);
+    }
+    return arr[i];
+  }
+
+  // Returns the length of the array
+  int getSize() { return len; }
```
Comment (Juliang0705) [inline initial-reverse/reverse.cc]: can mark the function const

**Review by Juliang0705 — COMMENTED** [review]
...

File: initial-reverse/reverse.cc (hunk 4 above)
Comment (BriungRi) [reply initial-reverse/reverse.cc]: What does having a const reference enforce? Is there an example of where I might want `get` to return a const reference?

Comment (Juliang0705) [reply initial-reverse/reverse.cc]: Returning pure T always makes a copy. Returning reference allow users to decide whether they want a copy or reference. "const" is optional but nice to have.

With ref, you can 
std::string a = vec.get(i) // gets a copy, a.k.a. another malloc 
const std::string & a = vec.get(i) // gets a reference, a.k.a pointer copy, no malloc

**Comment by BriungRi** [comment]
ty for the review @Juliang0705 ❤

**Pull request merged**